src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -401,56 +399,66 @@ where
 
         loop {
             tokio::select! {
-                message = source_stream.next() => match message {
-                    Some(SourceMessageType::Message(message, cap, diff)) => {
-                        let new_status = match message {
-                            Ok(_) => HealthStatus::Running,
-                            Err(ref error) => {
-                                HealthStatus::StalledWithError(error.inner.to_string())
+                _ = source_stream.as_mut().peek() => {
+                    let timer = Instant::now();
+
+                    // We want to efficiently batch up messages that are ready. To do that we will
+                    // activate the output handle here and then drain the currently available
+                    // messages until we either run out of messages or run out of time.
+                    let mut output = output.activate();
+                    while timer.elapsed() < YIELD_INTERVAL {
+                        match source_stream.next().now_or_never() {
+                            Some(Some(SourceMessageType::Message(message, cap, diff))) => {
+                                let new_status = match message {
+                                    Ok(_) => HealthStatus::Running,
+                                    Err(ref error) => {
+                                        HealthStatus::StalledWithError(error.inner.to_string())
+                                    }
+                                };
+
+                                let new_status_update = HealthStatusUpdate {
```

Comment:
side note (not for this PR): all this status handling per message seems a bit wasteful, but idk.. there's bigger fish to fry 🤷